src/Pinch/Internal/RPC.hs

```diff
@@ -0,0 +1,38 @@
+{-# LANGUAGE ExistentialQuantification  #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TypeFamilies               #-}
+module Pinch.Internal.RPC
+  ( Channel(..)
+  , createChannel
+  , createChannel1
+  ) where
+
+
+import           Data.Hashable        (Hashable)
+import           Data.Text            (Text)
+
+import qualified Data.HashMap.Strict  as HM
+import qualified Data.Text            as T
+
+import           Pinch.Internal.TType
+import           Pinch.Internal.Value
+import           Pinch.Protocol       (Protocol)
+import           Pinch.Transport      (Connection, Transport)
+
+-- | A bi-directional channel to read/write Thrift messages.
+data Channel = Channel
+  { cTransportIn  :: !Transport
+  , cTransportOut :: !Transport
+  , cProtocolIn   :: !Protocol
+  , cProtocolOut  :: !Protocol
+  }
+
```

Comment:
I added a Message-based API to channel, code-wise it doesn't make a huge difference but I think conceptually it fits nicely.

An example of what the code-gen currently produces (incompatible with current pinch):

https://gist.github.com/phile314/8f0bcc456b236430df2700c5bbbb2f6f

Reply:
Thanks, the example helps.

To use the `_Args` struct in the client to construct the messages, the system will need a way to determine the Thrift method name from an `_Args` struct.  This could take the form of something like the following:

```
class (Pinchable a, Tag a => TStruct) => Request a where
  methodName :: Proxy a -> Text
  serviceName :: Proxy a -> Text
```

(The service name would be needed to implement a multiplexed Thrift server/client. In those cases, the method name sent over the wire is in the form `Trivial::success`.)

And I agree that accepting multiple arguments instead of the Args struct would be nicer, since the client code already does that.